src/snarky-typescript/src/index.ts

```diff
@@ -0,0 +1,517 @@
+    x: Var,
+    y: Var
+  ): BoolVar<F, this, CircuitVar<F, this>>;
+
+  abstract assertEqual<Var extends AsFieldElements<CircuitVar<F, this>, Var>>(
+    this: this,
+    x: Var,
+    y: Var
+  ): void;
+}
+
+type StandardConstraints = 'multiply' | 'equal' | 'boolean';
+
+function test<
+  F extends FieldElement<F>,
+  C extends Circuit<F, StandardConstraints, undefined, F[]>
+>(Circuit: C): F[] {
+  /* Field interface test. */
+  const f1 = Circuit.Field.one;
+  const f2 = new Circuit.Field(1);
+  const f3 = new Circuit.Field('1');
+  const f4 = f1.mul(2);
+  const f5 = f4.mul(f2);
+  const f6 = Circuit.Field.mul(f2, f3).add(f5);
+
+  /* Circuit interface test. */
+  const knownPublicInput = new Circuit.Var(f6, {isPublic: true});
+  const unknownPublicInput = new Circuit.Var({isPublic: true});
+  const knownPrivateInput = new Circuit.Var(0);
+  const unknownPrivateInput = new Circuit.Var();
+
+  unknownPublicInput.store(() => {
+    const val = knownPublicInput.read().mul(knownPrivateInput.read());
+    console.log(val.toString());
+    return val;
+  });
+
+  Circuit.run(() => {
+    console.log(unknownPublicInput.read());
+  });
+
+  unknownPrivateInput.store(15);
+
+  const res = knownPublicInput
+    .inv()
+    .mul(unknownPublicInput)
+    .add(knownPrivateInput)
+    .sub(unknownPrivateInput);
+
+  Circuit.addConstraint('equal', res, res);
+
+  const eq = res.equals(unknownPrivateInput);
+
+  /* More complicated types. */
+
+  class ECPoint {
+    x: F;
+    y: F;
+    constructor(x: F, y: F) {
+      this.x = x;
+      this.y = y;
+    }
+    toFieldElements(this: this): F[] {
+      return [this.x, this.y];
+    }
+    ofFieldElements([x, y]: F[]): ECPoint {
+      return new ECPoint(x, y);
+    }
+  }
+
+  /* This ugly extends will be replaced with a proper CircuitData class for
+   * concrete proof system classes. */
+  class VarECPoint extends CircuitData<
+    F,
+    C,
+    CircuitVar<F, C>,
+    BoolVar<F, C, CircuitVar<F, C>>,
+    ECPoint
+  > {
+    x: CircuitVar<F, C>;
+    y: CircuitVar<F, C>;
+    constructor(x: CircuitVar<F, C>, y: CircuitVar<F, C>) {
+      super();
+      this.x = x;
+      this.y = y;
+    }
+
+    /* TODO: These should be implemented in the abstract class with a generic
+     * implementation over toFieldElements.. */
+
+    toFieldElements(this: this): CircuitVar<F, C>[] {
+      return [this.x, this.y];
+    }
+    ofFieldElements([x, y]: CircuitVar<F, C>[]): VarECPoint {
+      return new VarECPoint(x, y);
+    }
+    equals(this: this, that: this | ECPoint): BoolVar<F, C, CircuitVar<F, C>> {
+      return this.x.equals(that.x).and(this.y.equals(that.y));
+    }
+    assertEquals(this: this, that: this): void {
+      Circuit.Bool.assertTrue(this.equals(that));
+    }
+
+    if_(
+      this: this,
+      cond: BoolVar<F, C, CircuitVar<F, C>> | boolean,
+      that: this | ECPoint
+    ): VarECPoint {
+      let b: BoolVar<F, C, CircuitVar<F, C>>;
+      if (cond instanceof BoolVar) {
```

Comment:
Matthew let me know that this isn't intended to be something clients of the API ever have to do themselves, so it's not an issue here -- but I just want to call out explicitly, that we should probably rethink any design that ever would require users to use `instanceof` to accomplish their task.